### ISSUE TITLE
Clarify if forgiving-base64 is allowed

### ISSUE THREAD

**Issue by ShadowJonathan** [matrix-org/matrix-doc#3211]
Encountered in https://github.com/ruma/ruma/issues/591, `Construct` sent a base64 string in `signatures` that synapse accepted, but which is https://infra.spec.whatwg.org/#forgiving-base64-decode, arguably "non-compliant", yet accepted by https://github.com/matrix-org/python-unpaddedbase64.

Rust crate `base64` snagged on it, and gave a way out with a mention to above link about "forgiving base64", is this allowed per the matrix spec? If so, I think a comment should be made about that.

**Comment by ShadowJonathan** [comment]
https://github.com/ruma/ruma/issues/591 is about the base64 string `3UmJnEIzUr2xWyaUnJg5fXwRybwG5FVC6GqMHverEUn0ztuIsvVxX89JXX2pvdTsOBbLQx+4TVL02l4Cp5wPCm`, @gbsf had the following to say about it;

> @ShadowJonathan said:
> > this is legitimate base64 behaviour, is seems
> But is it? The input `3UmJnEIzUr2xWyaUnJg5fXwRybwG5FVC6GqMHverEUn0ztuIsvVxX89JXX2pvdTsOBbLQx+4TVL02l4Cp5wPCm` encodes an extra 4 bits at the end, so it should be either 65 bytes with `...mA=` or 64 bytes with `...g==`

The repair website https://base64.guru/tools/repair repairs the aforementioned base64 string with `...g==`, and [this SO question](https://stackoverflow.com/questions/2941995/python-ignore-incorrect-padding-error-when-base64-decoding) says to tack two `=` padding characters after the string.

Python `unpaddedbase64` accepts all suggested strings, only with `...mA=` it gives a different byte sequence. (this is normal, [this](https://cryptii.com/pipes/base64-to-hex) converter says the same)

Rust `base64` only accepts the `...g==` equivalent in this case, and rejects all others unless allowed with `decode_allow_trailing_bits`, which [explicitly mentions forgiving base64](https://docs.rs/base64/0.13.0/base64/struct.Config.html#method.decode_allow_trailing_bits), thus this issue.

**Comment by ShadowJonathan** [comment]
So after some more digging, i encountered that a `...g` equivalent (a "correctly" unpadded one) is accepted via `base64` + `STANDARD_NO_PADDING`. Initially i assumed this meant that synapse was chopping off padding haphazardly off of base64-encoded signatures, but @gbsf mentioned that this Construct (the originating server of this base64 signature) has a no-pad encoder, which means the `...m` is untouched.

TL;DR (sorry for the spam):  synapse tolerated slightly malformatted unpadded base64, should this be encoded in the spec? Or is this a synapse bug?

**Comment by jakobvarmose** [comment]
From https://matrix.org/docs/spec/appendices#unpadded-base64:
> When decoding Base64, implementations SHOULD accept input with or without padding characters wherever possible, to ensure maximum interoperability.

I think this should be clarified, so all implementations can do the same. Specifically:

- Is non-canonical/forgiving encoding allowed?
- Are `-_` allowed instead of `+/`?
- Is padding allowed?
- Are spaces or other characters allowed?

The most straightforward would be to just reject all non-canonical base64, but since synapse currently accepts a lot of non-standard base64 that might not be a good idea.
